internal/cluster/certificates.go

```diff
@@ -0,0 +1,371 @@
+func hashCert(certificate *x509.Certificate) string {
+	spkiHash := sha256.Sum256(certificate.RawSubjectPublicKeyInfo)
+	return "sha256:" + strings.ToLower(hex.EncodeToString(spkiHash[:]))
+}
+
+// AsSecret will convert a single certificate into a Kubernetes secret.
+func (c *Certificate) AsSecret(cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) *corev1.Secret {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Namespace,
+			Name:      fmt.Sprintf("%s-%s", cluster.Name, c.Name),
+			Labels: map[string]string{
+				clusterv1.MachineClusterLabelName: cluster.Name,
+			},
+		},
+		Data: map[string][]byte{
+			TLSKeyDataName: c.KeyPair.Key,
+			TLSCrtDataName: c.KeyPair.Cert,
+		},
+	}
+
+	if c.Generated {
+		secret.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: bootstrapv1.GroupVersion.String(),
+				Kind:       "KubeadmConfig",
+				Name:       config.Name,
+				UID:        config.UID,
+			},
+		}
+	}
+	return secret
+}
+
+// AsFiles converts a slice of certificates into bootstrap files.
+func (c Certificates) AsFiles() []bootstrapv1.File {
+	clusterCA := c.GetCertificateByName(ClusterCAName)
+	etcdCA := c.GetCertificateByName(EtcdCAName)
+	frontProxyCA := c.GetCertificateByName(FrontProxyCAName)
+	serviceAccountCA := c.GetCertificateByName(ServiceAccountName)
+
+	return []bootstrapv1.File{
+		{
+			Path:        "/etc/kubernetes/pki/ca.crt",
+			Owner:       rootOwnerValue,
+			Permissions: "0640",
+			Content:     string(clusterCA.Cert),
+		},
+		{
+			Path:        "/etc/kubernetes/pki/ca.key",
+			Owner:       rootOwnerValue,
+			Permissions: "0600",
+			Content:     string(clusterCA.Key),
+		},
+		{
+			Path:        "/etc/kubernetes/pki/etcd/ca.crt",
+			Owner:       rootOwnerValue,
+			Permissions: "0640",
+			Content:     string(etcdCA.Cert),
+		},
+		{
+			Path:        "/etc/kubernetes/pki/etcd/ca.key",
+			Owner:       rootOwnerValue,
+			Permissions: "0600",
+			Content:     string(etcdCA.Key),
+		},
+		{
+			Path:        "/etc/kubernetes/pki/front-proxy-ca.crt",
+			Owner:       rootOwnerValue,
+			Permissions: "0640",
+			Content:     string(frontProxyCA.Cert),
+		},
+		{
+			Path:        "/etc/kubernetes/pki/front-proxy-ca.key",
+			Owner:       rootOwnerValue,
+			Permissions: "0600",
+			Content:     string(frontProxyCA.Key),
+		},
+		{
+			Path:        "/etc/kubernetes/pki/sa.pub",
+			Owner:       rootOwnerValue,
+			Permissions: "0640",
+			Content:     string(serviceAccountCA.Cert),
+		},
+		{
+			Path:        "/etc/kubernetes/pki/sa.key",
+			Owner:       rootOwnerValue,
+			Permissions: "0600",
+			Content:     string(serviceAccountCA.Key),
+		},
+	}
+}
+
+// Validate checks that all KeyPairs are valid
+func (c Certificates) Validate() error {
+	for _, certificate := range c {
+		if !certificate.KeyPair.IsValid() {
+			return errors.Errorf("CA cert material is missing cert/key for %s", certificate.Name)
```

Comment:
Might not strictly be a CA (if it's the service account cert)

Reply:
Removing this function entirely in favor of `EnsureAllExist`